### PR TITLE
[Routing] fix matching host patterns, utf8 prefixes and non-capturing groups

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 4.1
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #27448, #27461, #27504, #27512
| License       | MIT
| Doc PR        | -

### CODE REVIEW

**Comment by rmsint** [comment]
I ran into the `preg_match(): Compilation failed: ... unmatched parentheses ...` error and tried this fix. I could narrow it down to a utf8 route:

```php
/**
 * @Route(
 *     "/file/{filename}",
 *     name = "file_download",
 *     requirements = { "name": ".+" },
 *     options = { "utf8": true }
 * )
 * @Method("GET")
 */
public function download($filename, LocalFileSystem $localFileSystem)
{
  // ..
}
```
Similar as https://symfony.com/blog/new-in-symfony-3-2-unicode-routing-support

As soon as I remove the `options = { "utf8": true }` from the route my tests pass again and the route matcher does not throw the `preg_match(): ... unmatched parentheses` anymore.

**Comment by nicolas-grekas** [comment]
@rmsint see #27504, that's the issue you're experiencing. I'll add a fix for it here soon.

**Comment by rmsint** [comment]
@nicolas-grekas yes exactly, I noticed you pointed there to this fix. Thanks!

File: src/Symfony/Component/Routing/Matcher/Dumper/StaticPrefixCollection.php [hunk 1]
```diff
@@ -187,6 +185,11 @@ private function getCommonPrefix(string $prefix, string $anotherPrefix): array
             }
         }
         restore_error_handler();
+        if ($i < $end && 0b10 === (\ord($prefix[$i]) >> 6) && preg_match('//u', $prefix.' '.$anotherPrefix)) {
```
Comment (fabpot) [inline src/Symfony/Component/Routing/Matcher/Dumper/StaticPrefixCollection.php]: I would add a comment here explaining what that snippet of code do.

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Routing/Matcher/Dumper/StaticPrefixCollection.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Routing/Matcher/Dumper/StaticPrefixCollection.php]: comment added

**Comment by nicolas-grekas** [comment]
Status: needs review

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by teohhanhui** [comment]
Shouldn't we consider using a proper parser for the Routing component? Similar to: https://github.com/rcs/route-parser

https://github.com/rcs/route-parser/blob/12f4077798fbeab39e1abb8b2a4ad2fac1919809/lib/route/compiled-grammar.js

**Comment by stof** [comment]
@teohhanhui we are already parsing routes properly, and compiling them to a regex each (which is what this route-parser is doing) is battle-tested in symfony since many years.

the hard thing here is that Symfony 4.1 is now generating only a few regexes combining routes together to make matching much faster. And the new logic is not yet as battle-tested for all edge-cases (and the fact that we combine routes creates much more weird cases than when matching them separately, as we need to think about how the cases interact together).
The route-parser you linked too will match a single route, not a full collection of routes to find the matching one.

**Comment by teohhanhui** [comment]
Sorry, I must have misunderstood. Because I cannot find any parser in the Symfony Routing component.

**Pull request merged**